src/Protocols.Abstractions/Features/IConnectionTransportFeature.cs

```diff
@@ -2,17 +2,24 @@
 using System.Collections.Generic;
 using System.IO.Pipelines;
 using System.Text;
+using System.Threading.Tasks;
 
 namespace Microsoft.AspNetCore.Protocols.Features
 {
     public interface IConnectionTransportFeature
     {
         PipeFactory PipeFactory { get; }
 
-        IPipeConnection Connection { get; set; }
+        IPipeConnection Transport { get; set; }
+
+        IPipeConnection Application { get; set; }
 
         IScheduler InputWriterScheduler { get; }
 
         IScheduler OutputReaderScheduler { get; }
+
+        Task ConnectionAborted { get; }
+
+        Task ConnectionClosed { get; }
```

Comment:
It's broken because of the dispatching. Ordering problems cause issues that I don't want to fix while changing everything else.